core/src/main/java/org/apache/calcite/sql/SqlSelect.java

```diff
@@ -288,4 +297,133 @@ public boolean hasWhere() {
   public boolean isKeywordPresent(SqlSelectKeyword targetKeyWord) {
     return getModifierNode(targetKeyWord) != null;
   }
+
+  /**
+   * Rewrite a select statement with a where condition if the where
+   * condition depends on the columns in the select list. This writes
+   * the code to use two select statements.
+   *
+   * For example <code>Select A + 1 as X from table where X > 3</code>
+   * becomes
+   * <code>Select TEMP_COLUMN0 FROM (Select A + 1 as TEMP_COLUMN0)</code>
+   * <code>where TEMP_COLUMN0 > 3</code>
+   *
+   * We only do this replacement if we detect that the where may reference an alias generated
+   * by the select statement. This approach simplifies code but is not required for correctness.
+   * If this fails then the alias will be inlined as a fallback in validation.
+   *
+   * @return The new top level select
+   */
+  public SqlSelect rewriteSqlSelectIfWhereAlias() {
+    // Right now to make sure everything executes in order we require
+    // a select statement only contain a where
+    if (!keywordList.isEmpty() || where == null || groupBy != null
+        || having != null || qualify != null || !windowDecls.isEmpty()
+        || orderBy != null || offset != null || fetch != null) {
+      // Only perform the rewrite if we just have a where.
+      return this;
+    }
+    // Create a set of Alias names that could be referenced by a where
+    Set<ImmutableList<String>> possibleAliases = new HashSet<>();
+    // Iterate through the select statements and check for any aliases that
+    // may be used in the where.
+    for (int i = 0; i < selectList.size(); i++) {
+      SqlNode selectVal = selectList.get(i);
+      if (selectVal instanceof SqlBasicCall && selectVal.getKind() == SqlKind.AS) {
+        SqlBasicCall innerAlias = (SqlBasicCall) selectVal;
+        // Check if the alias is unnecessary. If so skip the transformation.
+        List<SqlNode> operands = innerAlias.getOperandList();
+        SqlIdentifier aliasIdentifier = (SqlIdentifier) operands.get(operands.size() - 1);
+        if (operands.size() == 2 && (operands.get(0) instanceof SqlIdentifier
+            || operands.get(0) instanceof SqlLiteral)) {
+          // No need to replace if the alias is a simple column
+          // rename or a simple name for a literal.
+          continue;
+        }
+        possibleAliases.add(aliasIdentifier.names);
```

Comment:
I would check here if we see the alias Identifier in the input rows

Reply:
Could you clarify this comment. I'm not sure I understand the request.

Reply:
IE, check if this alias has the same identifier as one of the columns in the input table, and we can't perform this optimization due to the issue described above.